sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubIOJsonTable.java

```diff
@@ -0,0 +1,222 @@
+          @ProcessElement
+          public void processElement(ProcessContext context) {
+            PubsubMessage pubsubMessage = context.element();
+            long msSinceEpoch = asMsSinceEpoch(pubsubMessage.getAttribute(getTimestampAttribute()));
+            context.output(KV.of(new DateTime(msSinceEpoch), pubsubMessage));
+          }
+        });
+  }
+
+  private static long asMsSinceEpoch(String timestamp) {
+    try {
+      return Long.parseLong(timestamp);
+    } catch (IllegalArgumentException e1) {
+      return parseRfc3339(timestamp).getValue();
+    }
+  }
+
+  private ParDo.SingleOutput<KV<DateTime, PubsubMessage>, Row> parseJsonPayload() {
+    return ParDo.of(
+        new DoFn<KV<DateTime, PubsubMessage>, Row>() {
+          private transient volatile @Nullable ObjectMapper objectMapper;
+
+          @ProcessElement
+          public void processElement(ProcessContext context) {
+            DateTime timestamp = context.element().getKey();
+            byte[] payloadBytes = context.element().getValue().getPayload();
+            String payloadJson = new String(payloadBytes, StandardCharsets.UTF_8);
+            Row payloadRow = parseRow(payloadJson);
+
+            context.output(
+                Row
+                    .withSchema(getSchema())
+                    .addValues(payloadRow.getValues())
+                    .addValue(timestamp)
+                    .build());
+          }
+
+          private Row parseRow(String json) {
+            if (objectMapper == null) {
+              objectMapper = newObjectMapperWith(RowJsonDeserializer.forSchema(getPayloadSchema()));
+            }
+
+            try {
+              return objectMapper.readValue(json, Row.class);
+            } catch (IOException e) {
+              throw new IllegalArgumentException("Unable to parse Json: " + json, e);
+            }
+          }
+        });
+  }
+
+  @Override
+  public PTransform<? super PCollection<Row>, POutput> buildIOWriter() {
+    throw new UnsupportedOperationException("Writing to a Pubsub topic is not supported");
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setPayloadSchema(Schema payloadSchema);
+    abstract Builder setSchema(Schema schema);
+    abstract Builder setTimestampAttribute(String timestampAttribute);
+    abstract Builder setTopic(String topic);
+
+    abstract PubsubIOJsonTable autoBuild();
+
+    abstract Schema getPayloadSchema();
+    abstract String getTimestampAttribute();
+
+    public PubsubIOJsonTable build() {
+      return
+          this
+              .setSchema(addTimestampField(getPayloadSchema(), getTimestampAttribute()))
+              .autoBuild();
+    }
+
+    private Schema addTimestampField(Schema schema, String timestampField) {
+      return
+          Schema
+              .builder()
+              .addFields(schema.getFields())
+              .addField(Schema.Field.of(timestampField, DATETIME.type()))
```

Comment:
This is going to be confusing to users. They defined one schema, but the table actually has a different schema. `SELECT *` won't work as expected.

Reply:
I agree. This is just setting up the framework, extra work is needed to support timestamps better. In the doc, for example, the row representing the payload json is nested within the table definition, and the attributes and timestamp are metadata in the top level fields.

At the moment though we can't define rows in ddl, so this is a temporary solution to have some basic cases working.